lib/colors.nix

```diff
@@ -69,57 +80,171 @@ let
+            "green"
+            "orange"
+            "blue"
+            "magenta"
+            "cyan"
+          ];
+          # everything but first (black) and last (white) elements
+          ansi-list = compose [
+            lib.tail
+            lib.init
+          ];
+        in
+        builtins.listToAttrs (
+          lib.zipListsWith lib.nameValuePair (color-list ++ map (x: "bright-${x}") color-list) (
+            ansi-list ansi'.dark.toList ++ ansi-list ansi'.bright.toList
+          )
+        )
+        // {
+          yellow = palette.base0A;
+          brown = palette.base0F;
+        };
 
-      base-hex-bgr = lib.mapAttrs' (k: v:
-        let rgb = splitRGB v;
-        in lib.nameValuePair "${k}-hex-bgr" "${rgb.b}${rgb.g}${rgb.r}") base;
+      # paths like `hex`, `rgb.r` etc.
+      paths =
+        [
+          [ ]
+          [ "hex" ]
+          [
+            "hex"
+            "bgr"
+          ]
+        ]
+        ++ (lib.mapCartesianProduct
+          (
+            { fmt, chan }:
+            [
+              fmt
+              chan
+            ]
+          )
+          {
+            fmt = [
+              "hex"
+              "rgb"
+              "dec"
+            ];
+            chan = [
+              "r"
+              "g"
+              "b"
+            ];
+          }
+        );
 
-      based = base-hex // base-hex-rgb // base-rgb-rgb
-        // base-dec-rgb // base-hex-bgr // base-short
-        // mnemonic;
+      base-colors = lib.mapAttrs (lib.const _color) base;
+      # transforms a list of colors to a flat list expected by base16 templates
+      based = lib.concatMapAttrs (
+        name: value:
+        builtins.listToAttrs (
+          lib.forEach paths (path: {
+            name = lib.concatStringsSep "-" ([ name ] ++ path);
+            value = lib.getAttrFromPath path value;
+          })
+        )
+      );
 
-      toList = lib.attrValues base;
+      inherit (builtins) isList isAttrs;
 
-      withHashtag =
-        let hashtag = color: "#${color}";
-        in (lib.mapAttrs (_: hashtag) based) // { toList = map hashtag toList; };
+      /*
+        Maps a value by function `f`, recurring whenever `cond` is `true`
+        and the value under question is either attrs or list.
+      */
+      mapRecursiveCond =
+        cond: f:
+        lib.fix (
+          self: val:
+          if isAttrs val && cond val then
+            lib.mapAttrs (_: self) val
+          else if isList val && cond val then
+            map self val
+          else
+            f val
+        );
 
-    in based // { inherit toList withHashtag; };
+      extra =
+        palette:
+        mnemonic palette
+        // {
+          ansi = ansi palette;
+          toList = builtins.attrValues palette;
+        };
+      total = palette: based palette // extra palette;
+      total' =
+        palette:
+        mapRecursiveCond (x: lib.strings.isConvertibleWithToString x -> isList x) toString (total palette)
```

Comment:
Can you please explain what transformation does this line achieve?

Reply:
First of all it takes `total palette`; after that it transforms recursively:
1. If the value under question is not `toString`-able or is a list, recurse with `mapAttrs` or `map`
2. Otherwise use `toString` to turn it into a string

Effectively this transforms all color nodes into strings (they have a `__toString` attr so they are `toString`-able) while keeping the attrset (e.g. `mnemonic`, `ansi`, `ansi.bright`) and list (e.g. `ansi.dark.toList`) structure

Reply:
This looks like a workaround. I think you can remove both `__toString` and instead use a common attribute like `.asString`. Then it will look like `mapRecursiveCond (x: x ? asString) (x: x.asString) (total palette)`.

As far as I understood, you use the implicit conversion to string in only in one other place - in the code for the `_color` function. There you can do without this conversion: instead of `self.hex` just use the `hex` parameter. This will also make for a simpler control flow.

Reply:
My previous comment was incomplete: this transformation also affects numbers such as `base01-dec-r` and `base0E-rgb-b`

Reply:
Then maybe we can convert them to string in the _color function itself?

Reply:
They were specifically implemented in the way they currently are to make scripting easier; making all `_color` attributes strings would be against that intention

Reply:
A user might want to do calculations with colors (including but not limited to color space conversion) and parsing stringified numbers is suboptimal. On a second thought, the color constructor should perhaps be made public for this reason.

Reply:
Ok, but in your PR they still get converted to string, don't they? We can explicitly convert them to string as I proposed above, and provide additional attributes with the integer values.

Reply:
Not sure what you mean, they are converted to strings to match base16/24 spec, and they have the additional attributes in the `original` attrset.

Reply:
Oh, I got it. After another thought, the actual problem is the following. You open too much of the internals of the library (increasing the future maintenance burden) to save a little time for 1% of "power users".

But even further, even this potential time saving is under question. Right now, the features you've added are undocumented, so nobody will know about them, thus they are code bloat. Even if we add documentation, it would look like "remember that dec-r attribute? We called `lib.strings.toInt` for you and you can access this value via `original.base00.dec-r`", being another sort of bloat.

Reply:
In any case, as I've stated in the README, the primary purpose of base16.nix is to be an interface to the base16 standard. Any additional features I want to keep at minimum.

Reply:
BTW, can you please add the docs about the featuers you've added to the DOCUMENTATION.md?